### DISCUSSION TITLE
Introducing new Janet bindings, Janet-botan.

### DISCUSSION THREAD

**Discussion by tttuuu888** [janet-lang/janet#1433] [Show and tell]
Hi @bakpakin and all community members,
Thank you for creating and evolving this interesting language.
I liked Janet from the moment I saw it and immediately wanted to create something with it.
So I've been working on creating a new Janet binding for my favorite crypto library, Botan.
I hope this helps someone, and you can get more information from the link below.

https://tttuuu888.github.io/janet-botan/

I would appreciate any feedback you may have.

Thank you.

**Comment by sogaiu** [comment]
Haven't tried it out yet, but IIUC, one might be able to adapt the code in this [janet-botan test](https://github.com/tttuuu888/janet-botan/blob/d74eaf08942e22f5b664dde26509b9c35679ceec/test/test_totp.janet) to work with [Github's 2FA stuff](https://docs.github.com/en/authentication/securing-your-account-with-two-factor-authentication-2fa/configuring-two-factor-authentication#manually-configuring-a-totp-app).

May be I'll compare the output from adapted code with that of [oathtool](https://www.nongnu.org/oath-toolkit/oathtool.1.html) (^^;

---

_Update_: Seems to work :)

```janet
(import botan :as b)

# XXX: when using oathtool from command line for real use, don't type in key
#      as part of the invocation.  use - instead of the key and type the
#      key in after starting the invocation.  this is to prevent other
#      processes from snooping on the key value.
#
# $ oathtool --verbose \
#            --totp=SHA1 \
#            --digits=6 \
#            --time-step-size=30s \
#            --now="1970-01-01 00:00:59 UTC" \
#            3132333435363738393031323334353637383930
# Hex secret: 3132333435363738393031323334353637383930
# Base32 secret: GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ
# Digits: 6
# Window size: 0
# TOTP mode: SHA1
# Step size (seconds): 30
# Start time: 1970-01-01 00:00:00 UTC (0)
# Current time: 1970-01-01 00:00:59 UTC (59)
# Counter: 0x1 (1)
#
# 287082

# "12345678901234567890"
(def key (b/hex-decode "3132333435363738393031323334353637383930"))
(def digits 6)
(def timestep 30)

(def totp (b/totp/new key "SHA-1" digits timestep))

(:generate totp 59)
# =>
287082
```

---

Would be nice to see base32 encoding / decoding show up somewhere since Github seems to hand stuff out in base32.

May be since [spork has base64](https://github.com/janet-lang/spork/blob/5d52cb9d6a3756789c972bb091b387d15b9ac2a0/spork/base64.janet) already, adding base32 would make sense.

---

_Update 2_: Adapting spork's base64.janet might look like [this](https://github.com/sogaiu/janet-totp/blob/9554e0d5dbf4b9e1da2301e62ae5e8f5fc86efa1/janet-totp/base32.janet).
